# Pipeline for running PRScs-auto

<br>


## 1. Download and install PRScs

- Details for installation and required packages can be seen in [PRScs github page](https://github.com/getian107/PRScs)

- Download LD reference panel from the same page, using `1KG-EUR` and `UKBB-EUR`

<br>

## 2. Format your GWAS sumstats

> **Note:** Before running PRScs, you need to reformat your GWAS sumstats with the format as: `SNP A1 A2 BETA/OR P` 


You can reformat using the script <span style="color:red"> `toPRScsFormat.sh`</span> after filling the variables, specifically:
 
 - <span style="color:red"> `sumstats`</span>: Full path to your GWAS sumstats after format to PRScs.
 - <span style="color:red"> `headers`</span>: The names separated by comma for (SNP A1 A2 BETA/OR P) in your GWAS sumstats in the exactly same order.

<br>

Then you can run <span style="color:red"> `bash toPRScsFormat.sh`</span>.

<br>

## 3. Run PRScs-auto

You have to fill the variables in <span style="color:red"> `run_prscs_auto.sh`</span> which are described in [PRScs github page](https://github.com/getian107/PRScs).

> **Note:** You need to specify your own path to PRScs.py (<span style="color:red"> `PATH_TO_PRSCS_DIR`</span>), which is where you installed PRScs. For the <span style="color:red"> `chrom`</span> variable if you run batch jobs on cluster, please modify it accordingly to your cluster setting, e.g., $ARRAY_ID. If you cannot run batch jobs, you can use for loop to submit one job for one chromosome. 


Then you can run PRScs-auto using <span style="color:red"> `bash run_prscs_auto.sh`</span>. 

<br>

## 4. Calculate PRS using plink

Plink is activated to run `--score` using <span style="color:red"> `bash run_prs.sh`</span>, specifically:

 - <span style="color:red"> `bfile`</span>: Full path to plink files in .bed .bim .fam format
 - <span style="color:red"> `scorefile`</span>: Full path to the file generated by PRScs in **`Step 5`**.
 - <span style="color:red"> `outdir`</span>: Full path to your output *.profile files generated by plink.


<br>

## 5. Merge score files for each chromosome

> **Note:** If you generated *.profile files by each chromosome, which is recommended when you have a large target cohort, then you have to merge them first before calculating the accuracy of PRS.

You can merge the *.profile score files by running the below script after filling the variables, specifically:

 - <span style="color:red"> `wkdir`</span>: Full path to the directory with `*.profile*` score files generated by plink
 - <span style="color:red"> `file_pattern`</span>: a **regex pattern** in order to list score files for 22 chromosomes
 - <span style="color:red"> `outdir`</span>: Full path to your output with **`suffix .profile`** files for combined PRS

Then you can run <span style="color:red"> `bash merge_scoreFiles.sh`</span>.

<br>

## 6. Calculate the PRS accuracy metrics

Details have been given in **`run_create_prs_metrics.md`**.

<br>


<br>


